`-` 아래의 네트워크에서 옵티마이저를 adam, sgd를 선택하여 각각 적합시켜보고 testset의 loss를 성능비교를 하라. epoch은 5정도로 설정하라. 
```
net = tf.keras.Sequential()
net.add(tf.keras.layers.Flatten())
net.add(tf.keras.layers.Dense(50,activation='relu'))
net.add(tf.keras.layers.Dense(50,activation='relu'))
net.add(tf.keras.layers.Dense(10,activation='softmax'))
net.compile(optimizer=???,loss=tf.losses.categorical_crossentropy,metrics=['accuracy','Recall'])
```

In [1]:
import tensorflow as tf
import tensorflow.experimental.numpy as tnp

In [2]:
tnp.experimental_enable_numpy_behavior()

In [3]:
import matplotlib.pyplot as plt
import numpy as np

데이터는 MNIST data를 선택하도록 한다.

In [4]:
(x_train, y_train), (x_test, y_test) = tf.keras.datasets.fashion_mnist.load_data()

4431872/4422102 [==============================] - 0s 0us/step


In [5]:
X= x_train.reshape(-1,28,28,1)/255 
y = tf.keras.utils.to_categorical(y_train)
XX = x_test.reshape(-1,28,28,1)/255
yy = tf.keras.utils.to_categorical(y_test)

In [6]:
from tensorboard.plugins.hparams import api as hp

In [7]:
for o in ['adam','sgd']:
    logdir = 'logs/hphsy_{}'.format(o)
    with tf.summary.create_file_writer(logdir).as_default():
        net = tf.keras.Sequential()
        net.add(tf.keras.layers.Flatten())
        net.add(tf.keras.layers.Dense(50,activation='relu'))
        net.add(tf.keras.layers.Dense(50,activation='relu'))
        net.add(tf.keras.layers.Dense(10,activation='softmax'))
        net.compile(optimizer=o,loss=tf.losses.categorical_crossentropy,metrics=['accuracy','Recall'])
        cb3 = hp.KerasCallback(logdir, {'optimizer':o})
        net.fit(X,y,epochs=5,callbacks=cb3)
        _rslt=net.evaluate(XX,yy)[0] # loss
        tf.summary.scalar('loss', _rslt, step=1) 

Epoch 1/5
1875/1875 [==============================] - 8s 4ms/step - loss: 0.5237 - accuracy: 0.8155 - recall: 0.7546
Epoch 2/5
1875/1875 [==============================] - 6s 3ms/step - loss: 0.3919 - accuracy: 0.8590 - recall: 0.8288
Epoch 3/5
1875/1875 [==============================] - 6s 3ms/step - loss: 0.3544 - accuracy: 0.8698 - recall: 0.8450
Epoch 4/5
1875/1875 [==============================] - 6s 3ms/step - loss: 0.3297 - accuracy: 0.8796 - recall: 0.8571
Epoch 5/5
313/313 [==============================] - 1s 2ms/step - loss: 0.3792 - accuracy: 0.8624 - recall: 0.8379
Epoch 1/5
1875/1875 [==============================] - 5s 3ms/step - loss: 0.7866 - accuracy: 0.7329 - recall: 0.5778
Epoch 2/5
1875/1875 [==============================] - 5s 3ms/step - loss: 0.5089 - accuracy: 0.8221 - recall: 0.7659
Epoch 3/5
1875/1875 [==============================] - 5s 3ms/step - loss: 0.4633 - accuracy: 0.8388 - recall: 0.7951
Epoch 4/5
1875/1875 [==============================] - 5s 

In [8]:
%load_ext tensorboard

In [12]:
%tensorboard --logdir logs

`-` 결과적으로 loss의 측면에서 봤을 때 optimizer가 adam인 경우에는 loss=0.3792 이고 sgd인 경우에는 loss=0.4329 이므로 adam의 성능이 sgd보다 더 좋음을 확인할 수 있다.

In [11]:
!rm -rf logs
!kill 172